### PR TITLE
Adding support to enable resourceSpec

### LINKED ISSUE #872: Allow embedding specs in Pipeline and PipelineRun instead of referencing.
# Expected Behavior
I see that the `TaskRun` spec allow `Task`s and `PipelineResource`s to be embedded into the spec via the `taskSpec` and `resourceSpec` fields. ([Docs](https://github.com/tektoncd/pipeline/blob/master/docs/taskruns.md#example-with-embedded-specs), [Issue](https://github.com/tektoncd/pipeline/issues/313))  Creating just the `TaskRun` is simpler than creating all the CRDs and avoids the need to clean up ephemeral `Tasks` and `PipelineResources` that are only needed for the individual `TaskRun`. (ref https://github.com/tektoncd/pipeline/issues/544)

It would be great if `PipelineRun`s could expose a similar embedding mechanism so that the `Task`s, `Pipeline`, and `PipelineResource`s could all be embedded into the `PipelineRun` spec. Here is a simple example of what I would expect a fully embedded `PipelineRun` to look like (snipped together from other examples):
```yaml
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: do-the-thing
spec:
  resources:
    - name: source-repo
      resourceSpec:
        type: git
        params:
          - name: url
            value: https://github.com/pivotal-nader-ziada/gohelloworld
  pipelineSpec:
    resources:
      - name: source-repo
        type: git
    tasks:
      - name: skaffold-unit-tests
        taskSpec:
          inputs:
            resources:
              - name: workspace
                type: git
          steps:
            - name: build-and-push
              image: gcr.io/kaniko-project/executor
              command:
                - /kaniko/executor
              args:
                - --destination=gcr.io/my-project/gohelloworld
        resources:
          inputs:
            - name: workspace
              resource: source-repo
```
# Additional Info
The context for this is running ProwJobs as Tekton Pipelines.
Prow now has a controller that can run ProwJobs that define a `PipelineRun` spec, but currently the referenced `Pipeline` and `PipelineResource`s must already exist in the cluster. It would be neat if a ProwJob could specify all the needed information in the `PipelineRun` spec instead of having to reference other CRDs that must be separately managed by our automation.

### PR DESCRIPTION
**Pull request by pritidesai**
Its now possible to embed resourceSpec into PipelineRun, for example:

```yaml
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: pipelinerun-echo-greetings
spec:
  resources:
    - name: git-repo
        resourceRef:
          name: my-git-repo
```

Can be specified as:

```yaml
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
  name: pipelinerun-echo-greetings
spec:
  resources:
    - name: git-repo
      resourceSpec:
        type: git
        params:
          - name: url
            value: https://github.com/myrepo/myrepo.git
```

Partially Fixes #872 

# Changes

This PR has lot of changes :) and WIP, still adding integration test.   
First, added new fields in `PipelineRunSpec` (`pipelinerun_types.go`) and `PipelineResourceBinding` (`resource_types.go`). Followed by reading and resolving Inputs/Outputs to read specifications if specified. Also updated unit tests and added documentation.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) and [TaskRun](../tekton/publish-run.yaml) to build and release this image

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md)
are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes)
must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS)
and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes)
must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS),
and they must first be added
[in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

```

### CODE REVIEW

**Comment by bobcatfish** [comment]
@pritidesai this is very exciting!!! Thanks so much for working on this :D (And for including docs right off the bad!!) 

> Adding support to enable pipelineSpec and resourceSpec

Apologies for this request because it might be a bit frustrating to do, but would it be possible to separate embedding of the pipelineSpec and the resourceSpec into 2 separate PRs? I think we all agree that embedding resourceSpecs is something we want to do, but afaik we don't have total agreement about PipelineSpecs 🙏

**Comment by pritidesai** [comment]
hey @bobcatfish sure, I will create a separate PR 😢 what happened to adding `pipelineSpec` ?

**Comment by bobcatfish** [comment]
```
I0919 14:55:47.116] ---- Checking links in the markdown files ----
I0919 14:55:47.116] ----------------------------------------------
I0919 14:56:13.854] docs/pipelineruns.md
I0919 14:56:13.854] 	ERROR	https://www.apache.org/licenses/LICENSE-2.0
I0919 14:56:13.854] 		Dialing to the given TCP address timed out
I0919 14:56:14.629] --------------------------------------
```

hmmm weird

/test pull-tekton-pipeline-build-tests

**Comment by bobcatfish** [comment]
This is looking great, thanks for this @pritidesai !! And thanks for separating this from the pipeline spec changes :pray: 

/approve

Need at least one more owner to review (https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md):

/hold

File: examples/pipelineruns/pipelinerun-with-resourcespec.yaml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  serviceAccount: 'default'
+  resources:
+    - name: pipeline-git
+      resourceSpec:
+        type: git
+        params:
+          - name: revision
+            value: master
+          - name: url
+            value: https://github.com/tektoncd/pipeline
```
Comment (bobcatfish) [inline examples/pipelineruns/pipelinerun-with-resourcespec.yaml]: great example!

File: test/builder/task.go [hunk 2]
```diff
@@ -597,9 +597,6 @@ func TaskRunOutputsResource(name string, ops ...TaskResourceBindingOp) TaskRunOu
 	return func(i *v1alpha1.TaskRunOutputs) {
 		binding := &v1alpha1.TaskResourceBinding{
 			Name: name,
-			ResourceRef: v1alpha1.PipelineResourceRef{
-				Name: name,
-			},
```
Comment (bobcatfish) [inline test/builder/task.go]: thanks for fixing this!!! :pray:  pretty weird that we got away with it all this time :sweat_smile:

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 3]
```diff
@@ -533,24 +538,31 @@ func resolveConditionChecks(pt *v1alpha1.PipelineTask,
...
+			}
+			// Finally add it to the resolved resources map
+			rr[r.Name] = gotResource
+		} else if resourceBinding.ResourceSpec != nil && resourceBinding.ResourceSpec.Type != "" {
+			rr[r.Name] = &v1alpha1.PipelineResource{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: resourceBinding.Name,
+				},
+				Spec: *resourceBinding.ResourceSpec,
+			}
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go]: Note that in #1353 I'm refactoring the PipelineRun resource resolution logic so it will all happen in one place - then we should only need to do this check in function instead of in a few different places :D

File: pkg/reconciler/pipelinerun/resources/conditionresolution_test.go [hunk 4]
```diff
@@ -290,9 +290,6 @@ func TestResolvedConditionCheck_ToTaskResourceBindings(t *testing.T) {
 
 	expected := []v1alpha1.TaskResourceBinding{{
 		Name: "git-resource",
-		ResourceRef: v1alpha1.PipelineResourceRef{
```
Comment (dibyom) [inline pkg/reconciler/pipelinerun/resources/conditionresolution_test.go]: Why was this deleted?

Comment (bobcatfish) [reply pkg/reconciler/pipelinerun/resources/conditionresolution_test.go]: Sometimes you want to have a ResourceBinding without a Ref (e.g. if you're using a spec!). So in the tests for embedded TaskRun Resource specs, this was being set even tho it shouldn't be

(The other function actually doesn't set this)

(Might be worth adding this to the commit message if it isn't there already @pritidesai !)

**Comment by bobcatfish** [comment]
Looks like you hit the approve button @dibyom (vs. /approve) so imma merge!

/hold cancel

**Comment by bobcatfish** [comment]
I guess we'll just leave the commit message as is XD

**Comment by bobcatfish** [comment]
Looks like this might be a real failure @pritidesai !

```
sidecar_test.go:84: Creating Task "sidecar-test-task-1"
sidecar_test.go:89: Creating TaskRun "sidecar-test-task-run-1"
sidecar_test.go:99: Error waiting for TaskRun "sidecar-test-task-run-1" to finish: task run sidecar-test-task-run-1 failed!
sidecar_test.go:111: Error waiting for Pod "" to terminate both the primary and sidecar containers: resource name may not be empty
sidecar_test.go:116: Error getting TaskRun pod: resource name may not be empty
sidecar_test.go:140: Either the primary or sidecar containers did not terminate
```

But just in case:

/test pull-tekton-pipeline-integration-tests

**Comment by pritidesai** [comment]
> Looks like this might be a real failure @pritidesai !
> 
> ```
> sidecar_test.go:84: Creating Task "sidecar-test-task-1"
> sidecar_test.go:89: Creating TaskRun "sidecar-test-task-run-1"
> sidecar_test.go:99: Error waiting for TaskRun "sidecar-test-task-run-1" to finish: task run sidecar-test-task-run-1 failed!
> sidecar_test.go:111: Error waiting for Pod "" to terminate both the primary and sidecar containers: resource name may not be empty
> sidecar_test.go:116: Error getting TaskRun pod: resource name may not be empty
> sidecar_test.go:140: Either the primary or sidecar containers did not terminate
> ```
> 
> But just in case:
> 
> /test pull-tekton-pipeline-integration-tests

will check just in few, it has happened before where integration tests failed but then with `retest`, they succeeded, but will confirm.

File: pkg/reconciler/pipelinerun/resources/conditionresolution.go [hunk 5]
```diff
@@ -174,13 +174,22 @@ func (rcc *ResolvedConditionCheck) ToTaskResourceBindings() []v1alpha1.TaskResou
 	var trb []v1alpha1.TaskResourceBinding
 
 	for name, r := range rcc.ResolvedResources {
-		trb = append(trb, v1alpha1.TaskResourceBinding{
+		tr := v1alpha1.TaskResourceBinding{
 			Name: name,
-			ResourceRef: v1alpha1.PipelineResourceRef{
+		}
+		if r.SelfLink != "" {
```
Comment (bobcatfish) [inline pkg/reconciler/pipelinerun/resources/conditionresolution.go]: hey @pritidesai, somehow I didn't notice this when I looked before: what's the significance of `selfLink` here? it seems like its being used to decide if a resource should be embedded or not? 🤔

Comment (pritidesai) [reply pkg/reconciler/pipelinerun/resources/conditionresolution.go]: > hey @pritidesai, somehow I didn't notice this when I looked before: what's the significance of `selfLink` here? it seems like its being used to decide if a resource should be embedded or not? 🤔

Hey @bobcatfish yes `selfLink` is the deciding factor here and most reliable, if resource is specified through reference (`resourceRef`), `selfLink` will never be empty  and set to something like `/apis/tekton.dev/pipelineresources/my-git-resource` since `my-git-resource` is already created and can be referred to through this link. This was the best choice, as in `inputs` being type of `map[string]*v1alpha1.PipelineResource` in [input_output_steps.go line 56](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/pipelinerun/resources/input_output_steps.go#L56) always had `name` set to the resource name and `PipelineResource.name` set to the same name as resource name. Hope this helps.
